### ISSUE TITLE
Batched letterbox (aspect-ratio preserving pad) transform

### ISSUE THREAD

**Issue by RichardHtunn** [kornia/kornia#3872]
### 🚀 Feature Description

A native, GPU-accelerated letterbox transform (resizing + symmetric padding to strictly preserve aspect ratio) that operates directly on batched 4D tensors `[B, C, H, W]`.

### 📂 Feature Category

Geometry

### 💡 Motivation

Currently, if users want to resize and pad images to a square or specific rectangle while strictly preserving aspect ratio ("letterboxing"), they must manually calculate dynamic padding margins. 

When doing on-the-fly image preprocessing for GPU-resident training batches (e.g., Object Detection or OCR), falling back to CPU-bound loops using libraries like `cv2` creates a severe execution bottleneck. Having a native tensor operation in Kornia solves this problem directly on CUDA.

### 💭 Proposed Solution

Add a tensor-native transform in `kornia.geometry.transform` (e.g. `letterbox` or `pad_to_aspect_ratio`).

**Proposed API:**
```python
import torch
import kornia.geometry.transform as K

# x: [B, C, H, W] tensor
# target_size: (target_h, target_w)
output = K.letterbox(x, target_size=(512, 512), pad_value=114)

### 🔄 Alternatives Considered

- **OpenCV / PIL on CPU:** Looping over `cv2.resize` and `cv2.copyMakeBorder` on CPU, which breaks GPU-resident training workflows.
- **Manual dynamic padding:** Requiring users to compute padding margins manually in PyTorch and pass them to `kornia.geometry.transform.pad`, which adds boilerplate and risk of edge-case bugs (e.g., odd/even pixel distribution).

### 🎯 Use Cases

- **OCR Pipelines:** Where aspect ratio distortion severely reduces character recognition accuracy.
- **Object Detection (e.g., YOLO):** Standardizing input tensor dimensions without stretching object aspect ratios.
- **GPU-Resident Pipelines:** Accelerating on-the-fly batched tensor preprocessing directly on CUDA.

### 📝 Additional Context

I built and tested a standalone version for an OCR model training pipeline. On an Nvidia T4 GPU (Batch Size = 32, 1080p -> 512x512 target), native batched GPU letterboxing yielded a ~7.8x throughput speedup compared to sequential `cv2` calls (~10,600 FPS vs ~1,350 FPS).

I have unit tests handling edge cases (even vs. odd padding, rectangular targets) and am ready to adapt this to Kornia's codebase conventions.

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

**Comment by edgarriba** [comment]
we recently implemented a fused cpu/cuda kernel for this in the rust project and expose to python https://github.com/kornia/kornia-rs/blob/main/kornia-py/src/pipeline.rs

we should explore how to propelrly integrate with kornia/pytorch a the new Image/Tensor api is Dlpack device friendl fro zero copies

**Comment by RichardHtunn** [comment]
A fused Rust kernel with zero-copy DLPack integration sounds exactly like the right approach for maximum throughput.

I would love to help build out the PyTorch integration layer for this. I see that the new experimental kornia.image.Image API already includes support for creating images directly from DLPack capsules via Image.from_dlpack. Since the low-level logic is already handled in kornia-rs/pipeline.rs, I can focus on bridging it into kornia so it feels completely native to end users.

Could you point me to an existing file in the kornia codebase that serves as a good reference for how you are currently wrapping a kornia-rs DLPack pipeline into the new PyTorch Image API? Once I have a reference point, I can start drafting the Python-side integration for the letterbox transform!

**Comment by RichardHtunn** [comment]
Can I get any response?

**Comment by ducha-aiki** [comment]
@RichardHtunn Sorry for being slow.

Good references to start from:
 - kornia-rs: kornia-py/examples/preprocess_to_inference.py — this is the closest thing to what you're describing, and its 
fused CUDA kernel already does letterbox resize: NV12 → kornia_rs.cuda.Preprocessor (decode + letterbox + normalize + CHW  pack) → torch.from_dlpack → model, fully zero-copy and GPU-resident.
- kornia-rs: kornia-py/tests/test_torch_zero_copy.py and tests/test_dlpack.py — the zero-copy contract your integration tests  should assert (shared data_ptr(), mutation visibility).
- kornia: kornia/image/image.py (Image.from_dlpack / to_dlpack) — the kornia-side landing point; kornia/io/io.py shows the
current (numpy-copy) kornia_rs integration pattern the DLPack path would replace.

One note: kornia-py/src/pipeline.rs exposes the CPU fused pipeline through numpy, not DLPack — the zero-copy handoff lives on
the CUDA Preprocessor path, so that's the one to bridge.

That said, a pure-torch letterbox in kornia.geometry.transform is valuable on its own and doesn't need to wait on the rs bridge — happy to review that as a first PR, with the rs zero-copy path as a follow-up.

**Comment by edgarriba** [comment]
in principle the kornia-rs Image API support zero-copy via dlpack/cuda

**Comment by RichardHtunn** [comment]
Thanks for the guidance! I just opened a PR for Phase 1 (the pure-torch letterbox) here: https://github.com/kornia/kornia/pull/3936#issue-5112074077. Once that's reviewed, I'll dive into the zero-copy Rust stuff! @ducha-aiki  @edgarriba
